lib/srv/discovery/fetchers/eks.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/aws/aws-sdk-go/service/eks/eksiface"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	concurrencyLimit = 5
+)
+
+type eksFetcher struct {
+	EKSFetcherConfig
+}
+
+// EKSFetcherConfig configures the EKS fetcher.
+type EKSFetcherConfig struct {
+	// Client is the AWS eKS client.
+	Client eksiface.EKSAPI
+	// Region is the region where the clusters should be located.
+	Region string
+	// FilterLabels are the filter criteria.
+	FilterLabels types.Labels
+	// Log is the logger.
+	Log logrus.FieldLogger
+}
+
+// CheckAndSetDefaults validates and sets the defaults values.
+func (c *EKSFetcherConfig) CheckAndSetDefaults() error {
+	if c.Client == nil {
+		return trace.BadParameter("missing Client field")
+	}
+	if len(c.Region) == 0 {
+		return trace.BadParameter("missing Region field")
+	}
+
+	if len(c.FilterLabels) == 0 {
+		return trace.BadParameter("missing FilterLabels field")
+	}
+
+	if c.Log == nil {
+		c.Log = logrus.WithField(trace.Component, "fetcher:eks")
+	}
+	return nil
+}
+
+// NewEKSFetcher creates a new EKS fetcher configuration.
+func NewEKSFetcher(cfg EKSFetcherConfig) (Fetcher, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &eksFetcher{cfg}, nil
+}
+
+func (a *eksFetcher) Get(ctx context.Context) (types.ResourcesWithLabels, error) {
+	clusters, err := a.getEKSClusters(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return clusters.AsResources(), nil
+}
+
+func (a *eksFetcher) getEKSClusters(ctx context.Context) (types.KubeClusters, error) {
+	var (
+		clusters        types.KubeClusters
+		mu              sync.Mutex
+		group, groupCtx = errgroup.WithContext(ctx)
+	)
+	group.SetLimit(concurrencyLimit)
+
+	err := a.Client.ListClustersPagesWithContext(ctx,
+		&eks.ListClustersInput{
+			Include: nil, // For now we should only list EKS clusters
+		},
+		func(clustersList *eks.ListClustersOutput, _ bool) bool {
+			for i := 0; i < len(clustersList.Clusters); i++ {
+				clusterName := aws.StringValue(clustersList.Clusters[i])
+				// group.Go will block if the concurrency limit is reached.
+				// It will resume once any running function finishes.
+				group.Go(func() error {
+					cluster, err := a.getMatchingKubeCluster(groupCtx, clusterName)
+					// trace.CompareFailed is returned if the cluster did not match the matcher filtering labels
+					// or if the cluster is not yet active.
+					if trace.IsCompareFailed(err) {
+						a.Log.WithError(err).Debugf("Cluster %q did not match the filtering criteria.", clusterName)
+						// never return an error otherwise we will impact discovery process
+						return nil
+					} else if err != nil {
+						a.Log.WithError(err).Warnf("Failed to discover EKS cluster %q.", clusterName)
+						// never return an error otherwise we will impact discovery process
+						return nil
+					}
+
+					mu.Lock()
+					defer mu.Unlock()
+					clusters = append(clusters, cluster)
```

Comment:
The lock here isn't necessary, as AWS's paginated results require the previous page in order to get the next. To confirm:
https://github.com/aws/aws-sdk-go/blob/v1.44.114/service/eks/api.go#L2318
This runs the function, fetches the next result, runs the function on those results, etc.

Reply:
That is correct but we also fetch each cluster returned for the current page concurrently (`errgroup.Group`). In the end, we might have up to 5 (limited to 5 by `group.SetLimit(concurrencyLimit)`) concurrent go-routines trying to append data into `clusters` slice.